src/webui/api/torrentscontroller.cpp

```diff
@@ -1038,3 +1038,78 @@ void TorrentsController::categoriesAction()
 
     setResult(categories);
 }
+
+void TorrentsController::addTagsAction()
+{
+    checkParams({"hashes", "tags"});
+
+    const QStringList hashes {params()["hashes"].split('|')};
+    const QStringList tags {params()["tags"].split(',', QString::SkipEmptyParts)};
+
+    for (const auto &tagRaw : asConst(tags)) {
+        const QString tag {tagRaw.trimmed()};
+        applyToTorrents(hashes, [tag](BitTorrent::TorrentHandle *const torrent)
+        {
+            if (!torrent->hasTag(tag) && !torrent->addTag(tag))
+                throw APIError(APIErrorType::Conflict, tr("Incorrect tag name"));
+        });
+    }
+}
+
+void TorrentsController::removeTagsAction()
+{
+    checkParams({"hashes"});
+
+    const QStringList hashes {params()["hashes"].split('|')};
+    const QStringList tags {params()["tags"].split(',', QString::SkipEmptyParts)};
+
+    for (const auto &tagRaw : asConst(tags)) {
+        const QString tag {tagRaw.trimmed()};
+        applyToTorrents(hashes, [tag](BitTorrent::TorrentHandle *const torrent)
+        {
+            if (torrent->hasTag(tag) && !torrent->removeTag(tag))
+                throw APIError(APIErrorType::Conflict, tr("Incorrect tag name"));
+        });
+    }
+
+    if (tags.size() == 0) {
+        applyToTorrents(hashes, [](BitTorrent::TorrentHandle *const torrent)
+        {
+            torrent->removeAllTags();
+        });
+    }
+}
+
+void TorrentsController::createTagAction()
+{
+    checkParams({"tag"});
+
+    const QString tag {params()["tag"].trimmed()};
+
+    if (tag.isEmpty())
+        throw APIError(APIErrorType::BadParams, tr("Tag cannot be empty"));
+
+    if (!BitTorrent::Session::isValidTag(tag))
+        throw APIError(APIErrorType::Conflict, tr("Incorrect tag name"));
+
+    if (!BitTorrent::Session::instance()->addTag(tag))
```

Comment:
Just a note for other contributors.
Perhaps we should reconsider the indication of errors in the base classes. Just "return a bool" is extremely uncomfortable. Error codes or exceptions can greatly simplify their use.